src/main/java/com/example/demo/service/ProductService.java

```diff
@@ -22,93 +24,117 @@
+		this.productMapper = productMapper;
+		this.criteriaMapper = criteriaMapper;
+		this.attachmentMapper = attachmentMapper;
+		this.supplierRepository = supplierRepository;
+		this.categoryRepository = categoryRepository;
+		this.productRepository = productRepository;
+		this.attachmentRepository = attachmentRepository;
+	}
+
+	public List<ProductReadDto> viewProducts() {
+		var products = productRepository.findAll();
+		return StreamSupport.stream(products.spliterator(), false).map(productMapper::toDto).collect(Collectors.toList());
+	}
+
+
+	private void checkProductUniqueness(ProductDomain productDomain) {
+		var supplierDomain = supplierRepository.findById(productDomain.getSupplier().getId());
+		if (supplierDomain.isEmpty()) {
+			throw new NotFoundException("Supplier with id " + productDomain.getSupplier().getId() + " not found.");
+		}
+		var categoryDomain = categoryRepository.findById(productDomain.getCategory().getId());
+		if (categoryDomain.isEmpty()) {
+			throw new NotFoundException("Category with id " + productDomain.getCategory().getId() + " not found.");
+		}
+	}
+
+	public ProductReadDto addProduct(ProductCreateDto productCreateDto) {
+		var productDomain = productMapper.fromCreateDto(productCreateDto);
+		if (productDomain == null){
+			throw new MappingException("Product mapping error");
+		}
+		checkProductUniqueness(productDomain);
+		var check = productRepository.findByNameAndSupplierId(productDomain.getName(), productDomain.getSupplier().getId());
+		if (check.isPresent()) {
+			throw new AlreadyExistsException("Product with name " + productDomain.getName() + " and supplier id " + productDomain.getSupplier().getId() + " already exists");
+		}
+		return productMapper.toDto(productRepository.save(productDomain));
+
+	}
+
+	public ProductReadDto updateProduct(ProductUpdateDto productUpdateDto) {
+		var productDomain = productMapper.fromUpdateDto(productUpdateDto);
+		if (productDomain == null){
+			throw new MappingException("Product mapping error");
+		}
+		checkProductUniqueness(productDomain);
+		var existingProduct = productRepository.findById(productDomain.getId());
+		if (existingProduct.isEmpty()) {
+			throw  new NotFoundException("Product with id " + productDomain.getId() + " not found");
+		}
+		var check = productRepository.findByNameAndSupplierId(productDomain.getName(), productDomain.getSupplier().getId());
+		if (check.isPresent() && check.get().getId() != existingProduct.get().getId()) {
+			throw new AlreadyExistsException("Product with name " + productDomain.getName() + " and supplier id " + productDomain.getSupplier().getId() + " already exists");
+		}
+		var res = productRepository.save(productDomain);
+		return productMapper.toDto(res);
+	}
+
+	public void deleteProduct(UUID id) {
+		var product = productRepository.findById(id);
+		if (product.isEmpty()) {
+			throw new NotFoundException("Product with id " + id + " not found");
+		}
```

Comment:
This kind of logic is repeated several times in various places. Why can't you wrap it up into its own method and then call it where it is required?